File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int largestDiff = 0;
        int minSoFar = Integer.MAX_VALUE;
        
        for(int i = 0; i < prices.length; i++) {
            
            if(minSoFar > prices[i]) {
                minSoFar = prices[i];
            }
            else {
                largestDiff = Math.max(largestDiff, prices[i] - minSoFar);
            }
        }
        return largestDiff;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)

        if n < 2:
            return 0

        # Initialize variables
        minPrice = prices[0]
        maxProfit = 0

        # Iterate through the array
        for i in range(1, n):
            # Update minimum price
            minPrice = min(minPrice, prices[i])

            # Update max profit
            maxProfit = max(maxProfit, prices[i] - minPrice)

        return maxProfit